### PR TITLE
mat: Make factorization parameters part of the matrix type

### LINKED ISSUE #756: mat: consider reconciling option passing in factorise methods
At the moment there are three factorisation methods that take options, the methods on `Eigen`, `EigenSym`, `GSVD` and `SVD`. These take different approaches to specifying options for factorisation.

I'm not too sure whether these can be reconciled, or whether doing so would be a net benefit, but it is worth considering.

@btracey @vladimir-ch Views?

### PR DESCRIPTION
**Pull request by btracey**
Fixes #756.

Please take a look.

### CODE REVIEW

**Comment by btracey** [comment]
Partial implementation.

This is a possible fix for 756. I think the upside to this implementation is that it requires knowing very little about the decomposition -- it does the standard thing by default. For all the methods it is
```
var fa mat.Factorization
f.Factorize(a)
```
This gives a reasonable factorization for all the types, and if you want optimizations you can set the appropriate fields of the struct.

**Comment by kortschak** [comment]
This seems reasonable, but thinking.

**Comment by btracey** [comment]
Fixed up the remaining parts to this commit. I think moving the factors in is the way to go. Two outstanding points in my opinion:

1) It would be nice if the default behavior of Eigen be to compute the right eigenvectors and not the left eigenvectors, since that's typically what people want. This could be changed to be `Left` and `NoRight`, but I'm not sure it's worth the confusion.

2) I think this approach is clearer for SVD than the bit-fields approach. However, the bitfields approach is actually quite nice for GSVD, especially in the ease of saying GSVDNone. We could keep the bitfields around, add a Kind field, and then flip the meaning so that the default behavior is "All". I do like the reduction in API surface however, so it's not clear that it's worth it.

PTAL

File: mat/gsvd.go [hunk 1]
```diff
@@ -18,7 +18,17 @@ import (
 // variable×sample spaces to reduced and diagonalized "eigenvariable"×"eigensample"
 // spaces.
 type GSVD struct {
-	kind GSVDKind
+	// NoU specifies that the U singular vectors should be computed during
```
Comment (kortschak) [inline mat/gsvd.go]: "should *not*"

and below.

**Comment by btracey** [comment]
Replaced with #872

**Pull request closed without merging**